### PR TITLE
Added a component that adds mesh from a file

### PR DESCRIPTION
**Pull request by joshuahansel**
This does the following:

- Extracted intermediate class `GeneratedMeshComponent` out of `GeometricalComponent`
- Added `FileMeshComponent` as a new child of `GeometricalComponent`
- Extracted interface class `HeatStructureInterface` out of `HeatStructureBase` and inherited from it.
- Inherited `HeatStructureFromFile3D` from `FileMeshComponent` and `HeatStructureInterface`. `HeatStructureBase` is now for generated 2D heat structures only. This requires some checks/casts to change.

This fixes this inheritance:
```
Component -> GeometricalComponent -> Component2D -> HeatStructureBase -> HeatStructureFromFile3D
```
which is obviously wrong (and can be seen to be very wrong from all of the `suppressParameter` calls in its `validParams`). I plan to have an interface class `HeatStructureInterface` instead of the main line base class `HeatStructureBase`.

Now it's
```
Component -> GeometricalComponent -> FileMeshComponent -> HeatStructureFromFile3D
```

Refs #22354

### CODE REVIEW

**Comment by joshuahansel** [comment]
This is expected to break Sockeye. Here's the fix: https://hpcgitlab.inl.gov/idaholab/sockeye/-/merge_requests/555.

**Comment by loganharbour** [comment]
@joshuahansel category labels go on issues

File: modules/thermal_hydraulics/src/components/HeatStructureFromFile3D.C [hunk 1]
```diff
@@ -17,221 +16,63 @@ registerMooseObject("ThermalHydraulicsApp", HeatStructureFromFile3D);
...
-  params.set<std::vector<Real>>("widths") = {1};
-  params.suppressParameter<std::vector<Real>>("widths");
-  params.set<std::vector<unsigned int>>("n_part_elems") = {1};
-  params.suppressParameter<std::vector<unsigned int>>("n_part_elems");
-  params.addClassDescription(
-      "Heat structure component that can load the mesh from an ExodusII file");
+  InputParameters params = FileMeshComponent::validParams();
+  params += HeatStructureInterface::validParams();
+
+  params.addClassDescription("Heat structure component that loads the mesh from an ExodusII file");
```
Comment (GiudGiud) [inline modules/thermal_hydraulics/src/components/HeatStructureFromFile3D.C:22]:
```suggestion
  params.addClassDescription("Heat structure component that loads a 3D mesh from an ExodusII file");
```

**Review by GiudGiud — COMMENTED** [review]
havent looked at all the files yet

I think we may be able to couple already. I ll give it a try

File: modules/thermal_hydraulics/src/components/FileMeshComponent.C [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+
+    // Get the boundary ID and add the elem/side pair to the boundary
+    unsigned int bc_id = boundary_id_map[exio_helper.id_list[e]];
+    boundary_info.add_side(elem, mapped_side, bc_id);
+  }
+
+  // Generate and set the boundary name
+  for (const auto & id_and_name : new_ids_to_names)
+    boundary_info.sideset_name(id_and_name.first) = genName(_name, id_and_name.second);
+
```
Comment (GiudGiud) [inline modules/thermal_hydraulics/src/components/FileMeshComponent.C:197]: the nodeset names are missing. the nodesets are properly passed though

File: modules/thermal_hydraulics/include/components/GeneratedMeshComponent.h [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+private:
+  void generateNodeLocations();
+  unsigned int computeNumberOfNodes(unsigned int n_elems);
+  std::vector<Real> getUniformNodeLocations(Real length, unsigned int n_nodes);
+  void placeLocalNodeLocations(Real start_length,
+                               unsigned int start_node,
+                               std::vector<Real> & local_node_locations);
+
+public:
+  static InputParameters validParams();
```
[the remark this answers is not in the document]
Comment (joshuahansel) [reply modules/thermal_hydraulics/include/components/GeneratedMeshComponent.h:51]: THM has it at the bottom historically, but I've been moving it up in my new classes. Just forgot here - I'll change it.

File: modules/thermal_hydraulics/src/components/FileMeshComponent.C (hunk 2 above)
Comment (joshuahansel) [reply modules/thermal_hydraulics/src/components/FileMeshComponent.C:197]: I just tried adding this in this loop:
```
boundary_info.nodeset_name(id_and_name.first) = genName(_name, id_and_name.second);
```
but to no effect. The sidesets and nodesets are sharing the IDs, right?

Comment (GiudGiud) [reply modules/thermal_hydraulics/src/components/FileMeshComponent.C:197]: they should

Comment (joshuahansel) [reply modules/thermal_hydraulics/src/components/FileMeshComponent.C:197]: Yeah, the IDs in Paraview match. It appears to be not just a problem in this component, but other ones as well. And sometimes it's the other way around - the side sets being named as "surface_#". I'll look into this to see if I can see some kind of pattern...

Comment (joshuahansel) [reply modules/thermal_hydraulics/src/components/FileMeshComponent.C:197]: Ok, so apparently some gold files have "surface_#", but if I rerun those same tests, I get "nodelist_#" instead. And this appears to happen if we do
```
boundary_info.add_side(elem, side, id);
boundary_info.sideset_name(id) = name;
```
but not when we do
```
boundary_info.add_node(node, id);
boundary_info.nodeset_name(id) = name;
```

I found that this is what fixes it:
```
  for (const auto & id_and_name : new_ids_to_names)
  {
    const std::string sideset_name = genName(_name, id_and_name.second);
    boundary_info.sideset_name(id_and_name.first) = sideset_name;
    boundary_info.nodeset_name(id_and_name.first) = sideset_name;
  }

  boundary_info.build_node_list_from_side_list();
```
It doesn't seem to do anything if I omit the `nodeset_name` line, which surprised me.

`build_node_list_from_side_list()` is already getting executed elsewhere afterwards, so I don't know I need the extra call here, but whatever.

**Review by GiudGiud — APPROVED** [review]
works for heat conduction.

Cant get non-null coupling on INS or INSAD but it's unrelated (missing vector classes for second) reasons

We should work towards FE-FV coupling next

**Comment by GiudGiud** [comment]
Adding a valgrind recipe just in case

**Pull request merged**